azext_iot/central/providers/device_provider.py

```diff
@@ -0,0 +1,79 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from azext_iot.central import services as central_services
+
+
+class CentralDeviceProvider:
+    def __init__(self, cmd, app_id, token=None):
+        """
+        Provider for device/device_template APIs
+
+        Args:
+            cmd: command passed into az
+            app_id: name of app (used for forming request URL)
+            token: (OPTIONAL) authorization token to fetch device details from IoTC.
+                MUST INCLUDE type (e.g. 'SharedAccessToken ...', 'Bearer ...')
+                Useful in scenarios where user doesn't own the app
+                therefore AAD token won't work, but a SAS token generated by owner will
+        """
+        self._cmd = cmd
+        self._app_id = app_id
+        self._token = token
+        self._device_templates = {}
+        self._devices = {}
+
+    def get_device_template(
+        self, device_id, central_dns_suffix="azureiotcentral.com",
+    ):
+        device = self.get_device(device_id, central_dns_suffix)
+        device_template_urn = device["instanceOf"]
+
+        if not device_template_urn:
+            raise ValueError(
```

Comment:
It's good and pythonic that you raise exceptions. However from a CLI devEx perspective we don't want to show a stack trace to the user - which is what will happen with an uncaught exception. Using `CLIError` will hide the stack trace. You could catch the true underlying exception at a higher level and wrap it in a CLIError.

Reply:
There are various raised exceptions through-out this PR. Please review and apply the feedback.